library/common/extensions/filters/http/assertion/filter.cc

```diff
@@ -26,53 +28,91 @@ AssertionFilter::AssertionFilter(AssertionFilterConfigSharedPtr config) : config
+                                         nullptr, absl::nullopt, "");
+      return Http::FilterHeadersStatus::StopIteration;
+    }
   }
 
   return Http::FilterHeadersStatus::Continue;
 }
 
 Http::FilterDataStatus AssertionFilter::decodeData(Buffer::Instance& data, bool end_stream) {
   config_->rootMatcher().onRequestBody(data, statuses_);
-  if (!config_->rootMatcher().matchStatus(statuses_).matches_) {
+  auto& matchStatus = config_->rootMatcher().matchStatus(statuses_);
+  if (!matchStatus.matches_ && !matchStatus.might_change_status_) {
     decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
                                        "Request Body does not match configured expectations",
                                        nullptr, absl::nullopt, "");
     return Http::FilterDataStatus::StopIterationNoBuffer;
   }
 
   if (end_stream) {
-    decoder_callbacks_->sendLocalReply(Http::Code::OK, "Request Body match configured expectations",
-                                       nullptr, absl::nullopt, "");
-    return Http::FilterDataStatus::StopIterationNoBuffer;
+    auto empty_trailers = Http::RequestTrailerMapImpl::create();
+    config_->rootMatcher().onHttpRequestTrailers(*empty_trailers, statuses_);
+    auto& matchStatus = config_->rootMatcher().matchStatus(statuses_);
+    if (!matchStatus.matches_ && !matchStatus.might_change_status_) {
+      decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
+                                         "Request Trailers do not match configured expectations",
+                                         nullptr, absl::nullopt, "");
+      return Http::FilterDataStatus::StopIterationNoBuffer;
+    }
+    if (!matchStatus.matches_) {
+      decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
+                                         "Request Body does not match configured expectations",
+                                         nullptr, absl::nullopt, "");
+      return Http::FilterDataStatus::StopIterationNoBuffer;
+    }
   }
-
   return Http::FilterDataStatus::Continue;
 }
 
 Http::FilterTrailersStatus AssertionFilter::decodeTrailers(Http::RequestTrailerMap& trailers) {
   config_->rootMatcher().onHttpRequestTrailers(trailers, statuses_);
-  if (!config_->rootMatcher().matchStatus(statuses_).matches_) {
+  auto& matchStatus = config_->rootMatcher().matchStatus(statuses_);
+  if (!matchStatus.matches_ && !matchStatus.might_change_status_) {
     decoder_callbacks_->sendLocalReply(Http::Code::BadRequest,
                                        "Request Trailers do not match configured expectations",
                                        nullptr, absl::nullopt, "");
     return Http::FilterTrailersStatus::StopIteration;
   }
-
-  decoder_callbacks_->sendLocalReply(
-      Http::Code::OK, "Request Trailers match configured expectations", nullptr, absl::nullopt, "");
-  return Http::FilterTrailersStatus::StopIteration;
+  if (!matchStatus.matches_) {
```

Comment:
I think I am confused about why we would catch this here rather than in the decodeData call, is it because the matcher is not completely done with knowing that data _will not_ match until we see an end_stream == true in decodeData, or a decodeTrailers call (implying there is no more data to match?).

Same as above could we add comments so that the thinking is clear?